### DISCUSSION TITLE
Next Font Downloader version planning!

### DISCUSSION THREAD

**Discussion by GustavoPeredo** [GustavoPeredo/font-downloader#73] [General]
## Welcome 🎊

This discussion post is to discuss ideas, design, opinions and more regarding the next major version of Font Downloader

### What is this?

With libadwaita coming up, I want to focus my time on rewriting the entire app from the ground up in rust, and by doing that, solve some of the app's biggest designs flaws.

### Opinions from all sides are very welcome!

Please! Feel free to share your opinions regarding any topics of this discussion.

### Allow other font sources

Fixes #65 #48 and #59

Some fonts aren't distributed by Google Fonts and some people would rather not use Google services. The current app design makes it very hard to maintain up-to-date fonts.

To help this situation, I also created ![Open Font Repository](https://github.com/GustavoPeredo/open-font-repository), the idea is for it to host open source font files as well as a json file containing download urls.

### Migration to libadwaita

Actually fixes #63 #67 #68 #69 and #70

Migrating to libadwaita will allow widget recoloring, allowing the green theme to work for all users :D

### Replace webkit with opengl or vulkan font rendering

Fixes #49 and #46

By migrating to rust, I hope I can more easily use OpenGL or Vulkan to render font widgets, this will bring many side benefits as well, such as: Performance upgrade, lower RAM usage and font previews before clicking a font.

### UI Revamp

To accomodate Font Downloader's new features and keep it simple to use, a new design is needed. Below you can find a concept I made, but entire redesigns as well as feedbacks are very welcome.

![Font Downloader UI](https://user-images.githubusercontent.com/30641461/137582524-0d0df7d2-37ea-4c4f-8874-65e8cb42c5e1.png)

| Pros | Cons |
| --- | --- |
| Contains all features | Looks boring, too practical |

### Translations

After some reading and issue #58 , I'm planning on migrating to ![Weblate](https://hosted.weblate.org/). With this move, I hope there it becomes easier for future contributors to translate te and avoid #55 .

![Bildschirmfoto vom 2021-10-16 11-38-30](https://user-images.githubusercontent.com/30641461/137582776-15cd6e11-e31e-493d-9b35-e7e742d9539c.png)

Of course, if any contributers would like to express their opinion against this move, feel free to do so! Maybe there are reasons not to switch to weblate.

### Programming

The current plan is to transform Font Downloader in a front end for ![font-catcher](https://github.com/GustavoPeredo/Font-Catcher). This would technically allow for other Font Downloader forks using other platforms, such as ![Kirigami](https://develop.kde.org/frameworks/kirigami/) and ![MauiKit](https://mauikit.org/).

Currently the design is as follow:
![Font Downloader Flow](https://user-images.githubusercontent.com/30641461/137583052-f12b8ed2-8cf7-404c-b910-6b23cbe3ffc2.png)

Feedback is welcome!

## Sources ✏

The svg files for this post can be found on: !(Font Downloader Meta)[https://github.com/GustavoPeredo/Font-Downloader-Meta]

**Comment by GustavoPeredo** [comment]
In the current design, there is no "Developer Window", so @litch0 (#27), do you have any idea where it could be added?

**Reply by lmtr0** [reply]
I think It would be important to maintain the state of the fonts (for example, where it was downloaded). I think it would be simpler to start maintaining these states  in the `download fonts` step. Then every time you open the developer window you would just need to filter the state to discover the repository it is and display the information needed to, for example, add it to a web page.

Example:
- The font is downloaded from google fonts, then the developer window will use the template for google fonts. It would present a window to the user saying "Add this to the header of your page <link href="fonts.gstatic.com/..." rel="stylesheet">".

**Comment by lumingzh** [comment]
What about issue #79 ?

**Reply by lmtr0** [reply]
Interesting for sure, I like the idea of having the installed fonts on a separate page, it would also be cool if the app use the downloaded fonts if available instead of downloading them every time.

It would also be useful to implement a cache for the downloaded fonts. It happen to me today, I was using a font, and I accidentally opened a new one, then I click to go back, the problem is that I am on a super slow connection today, and therefore, it took a while to download the font again. So a session cache would be very welcome, It would also improve the speed of the app.

**Reply by GustavoPeredo** [reply]
> What about issue #79 ?

That seems like a good idea! The headerbar space would have better usage then, right now it is empty.

> if the app use the downloaded fonts if available instead of downloading them every time

I approve this change

**Comment by lmtr0** [comment]
The Cache:
![image](https://user-images.githubusercontent.com/57605930/137830480-2dc7396d-6beb-4dec-ac16-7466c184f51c.png)

**Reply by GustavoPeredo** [reply]
Would you rather have a system cache or a session cache? Another options is to download fonts to /tmp and leave the cleaning job to the OS

**Reply by lmtr0** [reply]
I think session cache would be better, but it can still be in tmp/

**Reply by GustavoPeredo** [reply]
Session cache it is!

**Comment by GustavoPeredo** [comment]
I won't be using Rust anymore for the GUI app... Having to recompile the gtk bindings everytime I build the app is turning into a pain... I'll create Font Catchers python bindings instead.

**Reply by lumingzh** [reply]
Maybe you should try to contact gtk-rs developers for advice.

**Reply by lumingzh** [reply]
https://github.com/gtk-rs

**Reply by lmtr0** [reply]
Yeah, gtk-rs builds once and is cached on my machine. What happen with yours?

**Reply by GustavoPeredo** [reply]
Thanks for the comments, they encouraged me to find a better fix for the problem. Rust version is alive!

**Comment by GustavoPeredo** [comment]
> Maybe you should try to contact gtk-rs developers for advice.

Good idea, I'll contact them first.

> Yeah, gtk-rs builds once and is cached on my machine. What happen with yours?

The 'flatpak-builder' command needs an empty directory to build, so whenever it builds the flatpak, it also deletes the build cache. That's the problem I'm fscing right now...

**Reply by lmtr0** [reply]
I got it, wouldn't it be better to just use cargo for development them.

Then when you finsihe, just for packing, use the flatpak builder

**Reply by GustavoPeredo** [reply]
> I got it, wouldn't it be better to just use cargo for development them.
> 
> Then when you finsihe, just for packing, use the flatpak builder

Yes, that could be a solution for flatpak (currently I'm compiling directly with system packages). But that would require splitting the packaging process into two parts (a cargo compilation command and binary packaging flatpak), I (personally) don't prefer doing that...

**Comment by GustavoPeredo** [comment]
It's alive!
![2021-11-23 11-18-35](https://user-images.githubusercontent.com/30641461/143007302-8c1c18dd-639d-4efb-8b25-bef3358d7a79.gif)

**Reply by lmtr0** [reply]
Nice

**Reply by GustavoPeredo** [reply]
> Nice

Thanks 😄
